### PR TITLE
feat(input): add help/descriptor text

### PR DESCRIPTION
**Pull request by stubbo**
Screenshot 2022-09-11 at 18 55 16

### CODE REVIEW

**Comment by stubbo** [comment]
@fullstackfool

File: src/components/input/Demo.vue [hunk 1]
```diff
@@ -200,6 +221,8 @@ export default defineComponent({
         const suffixProp = ref('100');
         const suffixSlot = ref('Feather-weight');
         const placeholder = ref<number|null>(null);
+        const helpProp = ref('I\'m an input!');
+        const helpSlot = ref('I\'m an input!');
```
Comment (nandi95) [inline src/components/input/Demo.vue:225]: One is enough here

Comment (nandi95) [reply src/components/input/Demo.vue:225]: Altough I realise no point of having v-models at all given the component isn't updated from the outside

File: src/components/input/Demo.vue [hunk 2]
```diff
@@ -175,6 +175,27 @@
...
                  placeholder="Placeholder text."
                  label="Placeholder" />
+
+        <UIInput v-model="helpProp"
+                 :error="error"
+                 :small="small"
+                 :loading="loading"
+                 name="ui-text15"
+                 class="mt-10"
+                 help="Idiot proof message..."
```
Comment (nandi95) [inline src/components/input/Demo.vue:185]: use some nicer message, only because this is open source 😅

File: src/components/input/UIInput.vue [hunk 3]
```diff
@@ -118,11 +118,19 @@
         </div>
 
         <UIExpandTransition>
-            <slot v-if="error || $slots.error" name="error">
-                <p class="text-color-error text-sm">
-                    {{ error }}
-                </p>
-            </slot>
+            <div v-if="error || $slots.error || help || $slots.help">
```
Comment (nandi95) [inline src/components/input/UIInput.vue:121]: you introduced a redundant `<div>` here. You can use `<template>` which is only a logical block and won't render anything.

File: src/components/input/UIInput.vue [hunk 4]
```diff
@@ -118,11 +118,19 @@
...
+                <slot v-if="error || $slots.error" name="error">
+                    <p class="text-color-error text-sm mt-1">
+                        {{ error }}
+                    </p>
+                </slot>
+
+                <slot v-if="(help || $slots.help) && !error && !$slots.error" name="help">
```
Comment (nandi95) [inline src/components/input/UIInput.vue:128]: this could be simplified by a `v-if=""` and `v-else-if=""`

File: src/components/input/UIInput.vue [hunk 5]
```diff
@@ -226,6 +234,10 @@ export default defineComponent({
             type: [Number, String]
         },
 
+        help: {
```
Comment (nandi95) [inline src/components/input/UIInput.vue:237]: Little comment here please like in the above props.

File: src/components/input/UIInput.vue [hunk 6]
```diff
@@ -118,11 +118,19 @@
...
+            <div v-if="error || $slots.error || help || $slots.help">
+                <slot v-if="error || $slots.error" name="error">
+                    <p class="text-color-error text-sm mt-1">
+                        {{ error }}
+                    </p>
+                </slot>
+
+                <slot v-if="(help || $slots.help) && !error && !$slots.error" name="help">
+                    <p class="text-color-muted text-sm mt-1">
+                        {{ help }}
```
Comment (nandi95) [inline src/components/input/UIInput.vue:130]: if we gonna have this, can you please implement the same for all other inputs too? Or at least the inputs which have the same box layout (textarea, select, file input)

Comment (nandi95) [inline src/components/input/UIInput.vue:130]: Could you add testing please? You'll find other examples how a prop/slot like this is tested

**Review by nandi95 — CHANGES_REQUESTED** [review]
🦦

File: src/components/input/UIInput.vue [hunk 7]
```diff
@@ -117,12 +117,20 @@
             </div>
         </div>
 
-        <UIExpandTransition>
-            <slot v-if="error || $slots.error" name="error">
-                <p class="text-color-error text-sm">
-                    {{ error }}
-                </p>
-            </slot>
+        <UIExpandTransition :height-duration="100" :opacity-duration="200">
```
Comment (nandi95) [inline src/components/input/UIInput.vue:120]: keep it default (as that's consistent with other inputs)

```suggestion
        <UIExpandTransition>
```

File: src/components/input/UIInput.test.ts [hunk 8]
```diff
@@ -320,4 +320,83 @@ describe('UIInput', () => {
...
+                modelValue: '',
+                name: 'input',
+                help
+            }
+        });
+
+        expect(wrapper.get('.ui-help-text').text()).toBe(help);
+    });
+
+    it('should display the only error text when both help and error are passed as prop', () => {
```
Comment (nandi95) [inline src/components/input/UIInput.test.ts:386]: 👍

File: src/components/input/Demo.vue [hunk 9]
```diff
@@ -175,6 +175,27 @@
...
+
+        <UIInput v-model="helpSlot"
+                 :error="error"
+                 :small="small"
+                 :loading="loading"
+                 name="ui-text15"
+                 class="mt-10"
+                 label="Help (slot)">
+            <template #help>
+                <span class="animate-pulse text-sm mt-1 text-color-muted">Descriptive message...</span>
```
Comment (nandi95) [inline src/components/input/Demo.vue:196]: no need for the pulse animation (bit distracting 😅), if you wanna sho it's custom just say `Slotted descriptive message...`

File: src/components/input/UIInput.vue [hunk 10]
```diff
@@ -117,12 +117,20 @@
...
+                        {{ error }}
+                    </p>
+                </slot>
+
+                <slot v-else-if="(help || $slots.help)" name="help">
+                    <p class="ui-help-text">
+                        {{ help }}
+                    </p>
+                </slot>
+            </template>
```
Comment (nandi95) [inline src/components/input/UIInput.vue:133]: sorry to be a pain but what do you think about this?
```suggestion
            <div v-if="error || $slots.error || help || $slots.help">
                <UIFadeTransition>
                    <slot v-if="error || $slots.error" name="error">
                        <p class="ui-error-text">
                            {{ error }}
                        </p>
                    </slot>

                    <slot v-else-if="(help || $slots.help)" name="help">
                        <p class="ui-help-text">
                            {{ help }}
                        </p>
                    </slot>
                </UIFadeTransition>
            </div>
```
cc @fullstackfool

**Review by nandi95 — CHANGES_REQUESTED** [review]
🦦

**Review by nandi95 — APPROVED** [review]
🦦

**Pull request merged**